autogen/agentchat/contrib/teachable_agent.py

```diff
@@ -0,0 +1,412 @@
+        self.user_comments = []  # Stores user comments until the end of each chat.
+
+    def close_db(self):
+        """Cleanly closes the memo store."""
+        self.memo_store.close()
+
+    def prepopulate_db(self):
+        """Initializes the DB with a few arbitrary memos."""
+        self.memo_store.prepopulate()
+
+    def _generate_teachable_assistant_reply(
+        self,
+        messages: Optional[List[Dict]] = None,
+        sender: Optional[Agent] = None,
+        config: Optional[Any] = None,  # Persistent state.
+    ) -> Tuple[bool, Union[str, Dict, None]]:
+        """
+        Generates a reply to the last user message, after querying the memo store for relevant information.
+        Uses self.analyzer to make decisions about memo storage and retrieval.
+        """
+        if self.llm_config is False:
+            return False, None  # Return if no LLM was provided.
+        if messages is None:
+            messages = self._oai_messages[sender]  # In case of a direct call.
+
+        # Get the last user message.
+        user_text = messages[-1]["content"]
+
+        # This is a normal user turn. Keep track of it for potential storage later.
+        self.user_comments.append(user_text)
+
+        if self.memo_store.last_memo_id > 0:
+            # Consider whether to retrieve something from the DB.
+            new_user_text = self.consider_memo_retrieval(user_text)
+            if new_user_text != user_text:
+                # Make a copy of the message list, and replace the last user message with the new one.
+                messages = messages.copy()
+                messages[-1]["content"] = new_user_text
+
+        ctxt = messages[-1].pop("context", None)  # This peels off any "context" message from the list.
+        msgs = self._oai_system_message + messages
+        response = oai.ChatCompletion.create(context=ctxt, messages=msgs, use_cache=self.use_cache, **self.llm_config)
+        response_text = oai.ChatCompletion.extract_text_or_function_call(response)[0]
+
+        return True, response_text
+
+    def learn_from_recent_user_comments(self):
+        """Reviews the user comments from the last chat, and decides what teachings to store as memos."""
+        print(colored("\nREVIEWING CHAT FOR USER TEACHINGS TO REMEMBER", "light_yellow"))
+        # Look at each user turn.
+        if len(self.user_comments) > 0:
+            for comment in self.user_comments:
+                # Consider whether to store something from this user turn in the DB.
+                self.consider_memo_storage(comment)
+        self.user_comments = []
+
+    def consider_memo_storage(self, comment):
+        """Decides whether to store something from one user comment in the DB."""
+        # Check for a problem-solution pair.
+        response = self.analyze(
+            comment,
+            "Does any part of the TEXT ask the agent to perform a task or solve a problem? Answer with just one word, yes or no.",
```

Comment:
What do you think of extract these prompts as const variable?

Reply:
That would definitely be more typical. On the other hand, I think it would make the code harder to understand. In the end (to silence my guilt) I told myself that natural language is the new programming language, at least in these cases.